File: src/main/java/com/fobgochod/git/commit/settings/GitCommitHelperConfigurable.java
package com.fobgochod.git.commit.settings;

import com.fobgochod.git.commit.domain.TypeItem;
import com.intellij.openapi.options.SearchableConfigurable;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.List;

/**
 * 这个类Settings 中的属性被创建的时候
 */
public class GitCommitHelperConfigurable implements SearchableConfigurable {

    private final GitCommitHelperState settings;
    private GitCommitHelperComponent component;


    public GitCommitHelperConfigurable() {
        settings = GitCommitHelperState.getInstance();
    }

    @NotNull
    @Override
    public String getId() {
        return "plugin.git.commit.helper";
    }


    @Nullable
    @Override
    public JComponent createComponent() {
        if (component == null) {
            component = new GitCommitHelperComponent();
        }
        return component.getMainPanel();
    }


    @Nullable
    @Override
    public String getHelpTopic() {
        return "help.git.commit.helper.configuration";
    }

    @Nls(capitalization = Nls.Capitalization.Title)
    @Override
    public String getDisplayName() {
        return "GitCommitHelper";
    }


    public void reset() {
        if (component != null) {
            component.reset(settings);
        }
    }

    @Override
    public boolean isModified() {
        if (component == null) {
            return false;
        }
        String template = component.getSettings().getTemplate().trim();
        List<TypeItem> typeItems = component.getSettings().getTypeItems();
        return !template.equals(settings.getTemplate()) || typeItems != settings.getTypeItems();
    }


    @Override
    public void apply() {
        if (component != null) {
            settings.setTemplate(component.getSettings().getTemplate());
            settings.setTypeItems(component.getSettings().getTypeItems());
        }
    }
}

File: src/main/kotlin/com/fobgochod/git/commit/settings/GitCommitHelperConfigurable.kt
package com.fobgochod.git.commit.settings

import com.fobgochod.git.GitBundle
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.options.SearchableConfigurable
import org.jetbrains.annotations.Nls
import javax.swing.JComponent


/**
 * 这个类Settings 中的属性被创建的时候
 */
class GitCommitHelperConfigurable : SearchableConfigurable {

    private var state: GitCommitHelperState = GitCommitHelperState.getInstance()
    private var component: GitCommitHelperComponent = GitCommitHelperComponent()

    override fun getId(): String {
        return "git.commit.helper"
    }

    override fun createComponent(): JComponent {
        return component.mainPanel
    }

    override fun getDisplayName(): @Nls(capitalization = Nls.Capitalization.Title) String {
        return GitBundle.message("plugin.name")
    }

    override fun reset() {
        component.typeTable.reset(state)
        ApplicationManager.getApplication().runWriteAction {
            component.templateEditor.document.setText(state.template)
        }
    }

    override fun isModified(): Boolean {
        val template = component.templateEditor.document.text.trim { it <= ' ' }
        val typeRows = component.typeTable.typeRows
        return template != state.template || typeRows != state.typeRows
    }

    override fun apply() {
        state.template = component.templateEditor.document.text
        state.typeRows = component.typeTable.typeRows
    }
}
